{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dc603d84-4a58-45d8-a36a-9821785bf04b",
   "metadata": {},
   "outputs": [],
   "source": [
    "## Manhattan plot using Bokeh\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from math import pi\n",
    "from bokeh.io import *\n",
    "from bokeh.plotting import *\n",
    "from bokeh.transform import *\n",
    "from bokeh.embed import *\n",
    "from bokeh.layouts import *\n",
    "from bokeh.models import * \n",
    "\n",
    "\n",
    "\n",
    "# View data using pandas\n",
    "df = pd.read_csv('.\\TSVs\\T1D_GWAS_add.tsv', sep='\\t')\n",
    "# Seperate by chromosome ID, and colour them\n",
    "index_cmap = linear_cmap('chr_id', palette = ['grey','black']*11,low=1,high=22)\n",
    "\n",
    "# Format figure\n",
    "p=figure(frame_width= 800, # graph size\n",
    "         plot_height=500, # graph size\n",
    "         title = None, # Title added in html\n",
    "         toolbar_location=\"right\", \n",
    "         tools=\"pan,hover,xwheel_zoom,zoom_out,box_zoom,reset,box_select,tap,undo,save\", # Allows mouse hover to bring up information\n",
    "         tooltips=\"@rsid: (@chr_id,@chr_pos)\" # Shows when mouse is hovered over plot\n",
    "         )\n",
    "\n",
    "# Create Manhattan Plot\n",
    "p.circle(x='cumulative_pos',y='-logp', # x,y for scatter graph\n",
    "        source=df, # Source of data from the tsv file\n",
    "        fill_alpha=0.8, # Thickness of line around plot \n",
    "        fill_color=index_cmap, # Colour of plot\n",
    "        size=6, #size of plot  \n",
    "        selection_color=\"red\",\n",
    "        hover_color=\"green\"\n",
    "        )\n",
    "\n",
    "p.xaxis.axis_label= 'Chromosome' # x-axis label \n",
    "p.yaxis.axis_label= '-logp' # y-axis label\n",
    "\n",
    "# Correct the x-axis by replacing the cumulative_pos with the Chromosome position\n",
    "p.xaxis.ticker = [119895261, 373943002, 537393504, 716119012, 834845071, 964538826.5, 1147957441, 1306654016, \n",
    "1396415411, 1540730893, 1674222993, 1823778230, 1930583064, 2055942929, 2141945578, 2202426536, 2302961360, 2388903150, \n",
    "2436333506.5, 2482024730, 2529377491, 2584586808]\n",
    "p.xaxis.major_label_overrides = {119895261:'1', 373943002:'2', 537393504: '3', 716119012:'4', 834845071:'5', 964538826.5:'6', 1147957441: '7', 1306654016: '8', \n",
    "1396415411:'9', 1540730893:'10', 1674222993:'11', 1823778230:'12', 1930583064:'13', 2055942929:'14', 2088496163 :'14', 2141945578:'15',\n",
    "2202426536:'16', 2302961360:'17', 2388903150:'18', 2436333506.5:'19', 2482024730:'20', 2529377491:'21', 2584586808: '22'}\n",
    "\n",
    "text_input = TextInput(value=\"default\", title=\"Label:\")\n",
    "text_input.js_on_change(\"value\", CustomJS(code=\"\"\"\n",
    "    console.log('text_input: value=' + this.value, this.toString())\n",
    "\"\"\"))\n",
    "\n",
    "show(text_input)\n",
    "show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd32ba16",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "### THE OLD MATPLOTLIB MANHATTAN PLOT\n",
    "\n",
    "#import pandas as pd\n",
    "#import numpy as np\n",
    "#import seaborn as sns\n",
    "#import matplotlib.pyplot as plt\n",
    "\n",
    "\n",
    "#df = pd.read_csv('.\\TSVs\\T1D_GWAS_add.tsv', sep='\\t')\n",
    "\n",
    "#my_data = df.sample(642)\n",
    "\n",
    "#g=sns.relplot(\n",
    "    #data = my_data, # for whole genome this number will be 10000, or however many you want to see\n",
    "    #x= 'cumulative_pos',\n",
    "    #y= '-logp',\n",
    "    #aspect = 5, # size of graph\n",
    "    #hue = 'CHR_ID', # for whole genome this is the chromosome\n",
    "    #palette= ['grey','black'] * 11, #random colour scheme, can be changed to any colour(s)\n",
    "    #linewidth=0,\n",
    "    #size=6,\n",
    "    #legend=None\n",
    "    #)\n",
    "\n",
    "# Editing x-axis and making it neater\n",
    "#g.ax.set_xlabel('Chromosome') # x-axis label\n",
    "\n",
    "# Setting where the labels in the x-axis should be\n",
    "#g.ax.set_xticks(df.groupby('CHR_ID')['cumulative_pos'].median())\n",
    "#g.ax.set_xticklabels(df['CHR_ID'].unique())\n",
    "\n",
    "\n",
    "# Add title to graph\n",
    "#g.fig.suptitle('Manhatton Plot showing Association between SNPs and T1DM in GWAS')\n",
    "\n",
    "# Create annotations for each SNP point\n",
    "#annotations = my_data[my_data['-logp'] > 20].apply(lambda p : g.ax.annotate(p['SNPS'], (p['cumulative_pos'], p['-logp'])), axis=1).to_list()\n",
    "\n",
    "#adjust_text(annotations, arrowprops = {'arrowstyle': '->', 'color':'blue'})\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd490ca0",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "vscode": {
   "interpreter": {
    "hash": "d3979e0c8db77b3c34b8d304a6ee44c4275d54aacc4afce6d4328eb08cdec0d6"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
